external/js/cothority/lib/skipchain/skipchain.js

```diff
@@ -0,0 +1,217 @@
+//"use strict";
+
+const net = require("../net");
+const protobuf = require("../protobuf");
+const misc = require("../misc");
+const identity = require("../identity.js");
+
+const kyber = require("@dedis/kyber-js");
+
+const co = require("co");
+
+const skipchainID = "Skipchain";
+
+class Client {
+  /**
+   * Returns a new skipchain client from a roster
+   *
+   * @param {cothority.Roster} roster the roster over which the client can talk to
+   * @param {string} last know skipblock ID in hexadecimal format
+   * @returns {SkipchainClient} A client that can talks to the skipchain services
+   */
+  constructor(group, roster, lastID) {
+    this.lastRoster = roster;
+    this.lastID = misc.hexToUint8Array(lastID);
+    this.group = group;
+  }
+
+  /**
+   * Returns the latest known skipblockID
+   *
+   * @returns {string} hexadecimal encoded skipblockID
+   */
+  get latestID() {
+    return misc.uint8ArrayToHex(this.lastID);
+  }
+
+  /**
+   * updateChain asks for the latest block of the skipchain with all intermediate blocks.
+   * It automatically verifies the transition from the last known skipblock ID to the
+   * latest one returned. It also automatically remembers the latest good known
+   * roster from the latest block.
+   * @return {Promise} A promise which resolves with the latest skipblock if
+   * all checks pass.
+   */
+  getLatestBlock() {
+    var fn = co.wrap(function*(client) {
+      const requestStr = "GetUpdateChain";
+      const responseStr = "GetUpdateChainReply";
+      const request = { latestId: client.lastID };
+      // XXX  somewhat hackyish but sets a realistic upper bound
+      const initLength = client.lastRoster.length;
+      var nbErr = 0;
+      while (nbErr < initLength) {
+        // fetches the data with the current roster
+        client.socket = new net.RosterSocket(client.lastRoster, skipchainID);
+        var data = null;
+        try {
+          data = yield client.socket.send(requestStr, responseStr, request);
+        } catch (err) {
+          return Promise.reject(err);
+        }
+        // verifies it's all correct
+        var lastBlock, err;
+        [lastBlock, err] = client.verifyUpdateChainReply(data);
+        if (!err) {
+          // tries again with random conodes
+          nbErr++;
+        }
+        // update the roster
+        client.lastRoster = identity.Roster.fromProtobuf(lastBlock.roster);
+        client.lastID = lastBlock.hash;
+        // if there is nothing new stop !
+        if (!lastBlock.forward || lastBlock.forward.length == 0) {
```

Comment:
@jeffallen See here. I think that's what I do. I take the latest roster available. See if there is any forward link. If there are any left I continue to contact the cothority with the latest roster to get updates.